src/extensions/nuget/Microsoft.DotNet.UpgradeAssistant.Extensions.NuGet/NuGetTransitiveDependencyIdentifier.cs

```diff
@@ -25,10 +28,18 @@ namespace Microsoft.DotNet.UpgradeAssistant.Extensions.NuGet
 {
     public class NuGetTransitiveDependencyIdentifier : ITransitiveDependencyIdentifier
     {
+        private const string UniquePathPart1 = "dotnet-ua";
+        private const string UniquePathPart2 = "restores";
+        private const string UniquePathFilename = "project.txt";
+
+        private static readonly ConcurrentDictionary<string, LibraryDependency> _packageDependencies = new();
```

Comment:
I think it may be easier to test/reason about to have a second implementation of `ITransitiveDependencyIdentifier` that provides the caching behavior given an original `ITransitiveDependencyIdentifier`. Then if we figure out a better way to get the data, the caching is still available to be done

That would make the registration something like this:

```csharp
services.AddSingleton(new CachingTransitiveDependencyIdentifier(new NuGetTransitiveDependencyIdentifier));
```

Also, probably shouldn't be `static` - being a singleton would hold the necessary data

Reply:
It occurs to me that the logic using this (line 89) doesn't take into account the version number of the package being looked up as part of its caching strategy. Fixing this now.

To address the injection vs static concern my thought is this: this cache _should_ be static and close to the behavior because we need not compute the dependencies of a particular nuget pkg & version repeatedly; injecting this behavior doesn't really buy us anything and makes debugging harder.

Regarding a new type of `TransitiveDependencyIdentifier` which takes a `NuGetTransitiveDependencyIdentifier` I'm not following the benefit here either as, per Nikolche, the caching provider was purpose-built for this and should be reused across instances. I'm not sure what the actual object model would look like should we build a new Identifier which _takes_ the Nuget one but somehow utilizes Nuget's caching provider under the covers. Creating one that inherits from and overrides some behavior could make some sense, but again the bottom line was that we should've been creating a single provider to begin with.

Reply:
I'm referring to the packageDependencies cache (not the NuGet cache thing).

That does open a question as to what benefit is the `_packageDeendencies` doing for us? Do we now have effictively 2 caches? Also, should the NuGet cache instance be shared across *all* usage?

Reply:
I was merely trying to avoid having to `new()` up the exact same information every time we ask for pkg info. When I did profiling, these kinds of objects were _numerous_ so I am trying to trim them down in an effort to optimize.

Reply:
Also, yes, [based on what @nkolev92 said about the caching provider](https://github.com/dotnet/upgrade-assistant/pull/1262#discussion_r951947404) :)